### PR TITLE
planner: add the basic functions of the functional dependency

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29766

Problem Summary:

### What is changed and how it works?

Introduce the basic functions of how to maintain the functional dependency.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/funcdep/fd_graph.go [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+	equiv  bool
+	// unique is flag used to distinguish unique lax FD and non-unique lax FD, the difference is as below:
+	// 1: for unique lax FD: the determinant's normal value can't be repeatable, only the null value can be multiple.
+	//		which means: only limitation of not-null attribute to its determinant side can strengthen it as strict one.
+	// 2: for non-unique lax FD: the determinant's normal value and null value can be repeatable.
+	// 		which means: we have to strength both side of not-null attribute to strengthen it as strict one.
+	uniLax bool
+}
+
+// FDSet is the main portal of functional dependency, it stores the relationship between (extended table/ physical table)'s
```
Comment (qw4990) [inline planner/funcdep/fd_graph.go:42]:
```suggestion
// FDSet is the main portal of functional dependency, it stores the relationship between (extended table / physical table)'s
```

File: planner/funcdep/fd_graph_test.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAddStrictFunctionalDependency(t *testing.T) {
+	fd := FDSet{
+		fdEdges: []*fdEdge{},
+	}
+	fe1 := &fdEdge{
+		from:   NewFastIntSet(1), // AB -> CDEFG
```
Comment (qw4990) [inline planner/funcdep/fd_graph_test.go:28]: Should be `NewFastIntSet(1, 2)` here?

File: planner/funcdep/fd_graph_test.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+		equiv:  false,
+	}
+	fe3 := &fdEdge{
+		from:   NewFastIntSet(3, 5), // CE -> B
+		to:     NewFastIntSet(2),
+		strict: true,
+		equiv:  false,
+	}
+	fd.fdEdges = append(fd.fdEdges, fe1, fe2, fe3)
+	res := fd.ReduceCols(NewFastIntSet(1, 2)).SortedArray()
```
[the remark this answers is not in the document]
Comment (winoros) [reply planner/funcdep/fd_graph_test.go:152]: Yes, this one is to check the following cases.
When there's `a->cd, c->de, ce->b` in the func dep set and we want to insert a func dep`ab->f`.
From the existing ones, we can get a->b. So the `ab->f` can be simplified as `a->f`.
So we can insert less functional dependencies than before.

File: planner/funcdep/fd_graph_test.go (hunk 2 above)
Comment (winoros) [reply planner/funcdep/fd_graph_test.go]: Yeah, let me fix it.

**Review by qw4990 — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]
LGTM

**Pull request merged**